pkg/services/dashboards/database/database.go

```diff
@@ -994,16 +997,220 @@ func (d *dashboardStore) FindDashboards(ctx context.Context, query *dashboards.F
+	Term        string `xorm:"term"`
+	FolderTitle string `xorm:"folder_title"`
+	FolderUID   string `xorm:"folder_uid"`
+}
+
+// searchData is an input data for the search query template below.
+type searchData struct {
+	SQLBuilder
+	Text    string
+	OrgID   int64
+	UserID  int64
+	TeamIDs []int64
+	Role    string
+	Limit   int64
+
+	Dialect        migrator.Dialect
+	HasSplitScopes bool
+}
+
+// SQLBuilder is a more-or-less reusable SQL query builder that is based on
+// text/template. SQL is not a composable language really, so having all
+// subqueries/partials/filters in one place helps to better understand the whole
+// query.
+// SQLBuilder is expected to be embedded into the data struct passed into the
+// template. In a template one can use {{.Arg .Foo}} and it will replace that
+// with a "?" placeholder appending the value of .Foo to the Params slice.
+type SQLBuilder struct {
+	Params []any
+}
+
+func (sb *SQLBuilder) Arg(v any) (sql string) { sb.Params = append(sb.Params, v); return "?" }
+
+var searchSQL = template.Must(template.New("").Parse(`
+{{- /*
+####
+#### Here come different implementations for the innermost full-text search query.
+#### A query is expected to return a list of entities with their (OrgID, UID, Kind, Title, ParentFolderUID)
+####
+*/ -}}
+{{- define "text-search" -}}
+SELECT org_id, 'dashboard' as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }} AND NOT is_folder
+UNION ALL
+SELECT org_id, 'folder' as kind, uid, 0, parent_uid, title FROM folder WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }}
+{{ end }}
+{{- define "text-search-mysql-fts" -}}
+SELECT org_id, 'dashboard' as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE MATCH(title) AGAINST({{ .Arg .Text }}) AND NOT is_folder
+UNION ALL
+SELECT org_id, 'folder' as kind, uid, 0, parent_uid, title FROM folder WHERE MATCH(title) AGAINST({{ .Arg .Text }})
+{{ end }}
+{{- define "text-search-legacy" -}}
+SELECT org_id, 'dashboard' as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }} AND NOT is_folder
+UNION ALL
+SELECT org_id, 'folder' as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }} AND is_folder
+{{ end }}
+
+SELECT DISTINCT
+	entity.org_id, entity.kind, entity.uid, entity.id, entity.title,
+	dashboard_tag.term as term, f1.uid as folder_uid, f1.title as folder_title FROM (
+	{{- template "text-search-legacy" . -}}
+) AS entity
+LEFT JOIN folder f1 ON (entity.parent_uid = f1.uid AND entity.org_id = f1.org_id)
+{{ if ne .UserID 0 }}   {{- /* If a user is not admin: need to filter by permissions */ -}}
+LEFT JOIN folder f2 ON (f1.parent_uid     = f2.uid AND f1.org_id     = f2.org_id)
+LEFT JOIN folder f3 ON (f2.parent_uid     = f3.uid AND f2.org_id     = f3.org_id)
+LEFT JOIN folder f4 ON (f3.parent_uid     = f4.uid AND f3.org_id     = f4.org_id)
```

Comment:
This is hardcoding the max nesting level in the query. Which is probably fine, but we should make a note of it somewhere (or maybe make it dynamic when constructing the query?).

Reply:
Looking at where Nested Folders development is heading towarts, I think it's safe to assume the depth is always bounded. We can of course replace with with a loop to generate it, but for a depth of 4 I find it more readable with a query like this.